[llvm-objdump] Fix --source with --macho flag

rjmansfield opened the pull request:
The --source option was broken when using the --macho flag because DisassembleMachO() only initialized debug info when UseDbg was true, and would return early if no dSYM was found.

rjmansfield committed:
[llvm-objdump] Fix --source with --macho flag

The --source option was broken when using the --macho flag because
DisassembleMachO() only initialized debug info when UseDbg was true,
and would return early if no dSYM was found.

jh7370 reviewed (commented):
I've added a couple of people who are more familiar with the details of Mach-O format. Hopefully one of them can review this for you.

drodriguez commented on llvm/tools/llvm-objdump/MachODump.cpp:
```diff
@@ -7415,18 +7416,24 @@ static void DisassembleMachO(StringRef Filename, MachOObjectFile *MachOOF,
+    if (!DSym)
+      return;
+    DbgObj = DSym;
+    if (UseDbg) {
```
Should `PrintLines` also be checked here?

The docs for `--line-numbers` say "When disassembling, display source line numbers.", while the docs for `-g` say "Print line information from debug info if available". It seems to be the same thing, except that `-g` is exclusive for the Mach-O parser. Probably they were both introduced without knowledge of each other.

rjmansfield replied:
Yes, good catch.

drodriguez commented on llvm/tools/llvm-objdump/MachODump.cpp:
```diff
@@ -7415,18 +7416,24 @@ static void DisassembleMachO(StringRef Filename, MachOObjectFile *MachOOF,
 
+  SourcePrinter SP(DbgObj, TheTarget->getName());
+  LiveElementPrinter LEP(*MRI, *STI);
+
```
Not very important, and I haven't double checked the code compiles/works as I think it should.

This would happen even if `PrintSource` or `PrintLines` is not used. Also `FOS` below is created for each line to be printed, and later destroyed (probably cheap, but…)

With a lambda, it is possible to avoid some of these problems:

```
std::function<void(object::SectionedAddress Address, StringRef ObjectFilename)> printSourceLineIfNeeded;
if (PrintSource || PrintLines) {
  printSourceLineIfNeeded = [DbgObj, TheTarget, &MRI, &STI](object::SectionedAddress Address, StringRef ObjectFilename) {
    static SourcePrinter SP(DbgObj, TheTarget->getName());
    static LiveElementPrinter LEP(*MRI, *STI);
    static formatted_raw_ostream FOS(outs());
    SP.printSourceLine(FOS, Address, ObjectFilename, LEP);
  }
} else {
  printSourceLineIfNeeded = [](object::SectionedAddress Address, StringRef ObjectFilename) {};
}
...
// Below, use the lambda
printSourceLineIfNeeded({PC, SectIdx}, Filename);
```

rjmansfield replied:
DisassembleMachO can be called here multiple times e.g. different slices of a universal binary. The statics in the lambda might be an issue. I can use a std::optional here to avoid the unnecessary initialization.

drodriguez reviewed (commented):
Not sure if for `llvm-objdump` the options that are not explicitly Mach-O should work with Mach-O. Or should work the same than when `--macho` is not provided. It seems that `--macho` might have been introduced at a time that a lot of functionality was not existing.

I have checked, to the best of my ability, if this would change anything for `llvm-otool`, which output probably should be compatible with Apple's `otool`, and I don't think `otool` options give access to the same functionality that `--source`, `--line-numbers`, or `-g`, so the changes in the code should not be hit by `otool`.

rjmansfield committed:
Address reviewer feedback.

Fix --line-numbers with --macho and add test.
Use std::optional to defer initialization of SourcePrinter/LiveElementPrinter.
